File: spring-pulsar-samples/java-sample/src/main/java/com/intuit/spring/pulsar/java/sample02/PulsarExceptionHandlers.java
package com.intuit.spring.pulsar.java.sample02;

import com.intuit.spring.pulsar.client.exceptions.PulsarConsumerExceptionHandlerFunction;
import com.intuit.spring.pulsar.client.exceptions.PulsarExceptionHandler;
import com.intuit.spring.pulsar.client.exceptions.PulsarExceptionHandlerClass;
import com.intuit.spring.pulsar.client.exceptions.PulsarProducerExceptionHandlerFunction;
import org.springframework.stereotype.Component;

@Component
@PulsarExceptionHandlerClass
public class PulsarExceptionHandlers {

    @PulsarProducerExceptionHandlerFunction(exceptions = {ShortMessageProducerException.class})
    PulsarExceptionHandler producerExceptionHandler = exceptionHandlerParams -> {
        System.out.println("Exception occurred while performing " + exceptionHandlerParams.getAction());
        System.out.println("Handling producer exception " + exceptionHandlerParams.getException());
    };

    @PulsarConsumerExceptionHandlerFunction(exceptions = {LongMessageConsumerException.class})
    PulsarExceptionHandler consumerExceptionHandler = exceptionHandlerParams -> {
        System.out.println("Exception occurred while performing " + exceptionHandlerParams.getAction());
        System.out.println("Handling consumer exception " + exceptionHandlerParams.getException());
    };
}

File: spring-pulsar-samples/kotlin-sample/src/main/kotlin/com/intuit/spring/pulsar/kotlin/sample02/PulsarExceptionHandlers.kt
package com.intuit.spring.pulsar.kotlin.sample02

import com.intuit.spring.pulsar.client.exceptions.PulsarConsumerExceptionHandlerFunction
import com.intuit.spring.pulsar.client.exceptions.PulsarExceptionHandler
import com.intuit.spring.pulsar.client.exceptions.PulsarExceptionHandlerClass
import com.intuit.spring.pulsar.client.exceptions.PulsarProducerExceptionHandlerFunction
import org.springframework.stereotype.Component

/**
 * This class will add exception handlers to handle all exceptions
 * thrown by Pulsar producers and consumers
 */
@PulsarExceptionHandlerClass
@Component
class PulsarExceptionHandlers {
    @PulsarProducerExceptionHandlerFunction(ShortMessageProducerException::class)
    var pulsarProducerExceptionHandler = PulsarExceptionHandler { exceptionHandlerParams ->
        println("Exception occurred while performing ${exceptionHandlerParams.action}")
        println("Handling producer exception ${exceptionHandlerParams.exception}")
    }

    @PulsarConsumerExceptionHandlerFunction(LongMessageConsumerException::class)
    var pulsarConsumerExceptionHandler = PulsarExceptionHandler { exceptionHandlerParams ->
        println("Exception occurred while performing ${exceptionHandlerParams.action}")
        println("Handling consumer exception ${exceptionHandlerParams.exception}")
    }
}
